### DISCUSSION TITLE
bpftrace math library interest?

### DISCUSSION THREAD

**Discussion by janet-campbell** [bpftrace/bpftrace#3213] [Ideas]
Hi - I'm relatively new to bpftrace but not to kernel instrumentation.  I have found certain math functions to be useful, chief among those a high precision/partial power of 2 logarithm.  This lets you construct histograms and heatmaps with great detail and makes it easier to do math on data points in-kernel.

This is a latency heatmap of nfs_file_release across an HPC cluster.  bpftrace is doing an 8 divisions per power of 2 log (base=2^(1/8)):

<img width="1412" alt="Screenshot 2024-05-30 at 18 15 08" src="https://github.com/bpftrace/bpftrace/assets/171300685/620722ce-c7cd-4502-8624-4884806ce24a">

Is there appetite for a set of math library functions, and what would be helpful to people?  So far my priorities are logs/exp (using a Feynman/Knuth/Owen approach) and vector math, but trig should be pretty fast (using a CORDIC approach) if there's a use for it.

**Comment by danobi** [comment]
Yes, for sure we'd be interested. We're currently working on user-defined function support which would unlock a bpftrace "standard library" of sorts. Basically instead of writing all new functionality in LLVM IR, it'd let us write functionality in bpftrace scripting language (bpfscript) itself and ship that to users. 

I think a math library would be a prime candidate as first user. If we hit limitations, we can extend the core language as necessary.

**Reply by janet-campbell** [reply]
That's great.  Most of these should be quite fast if done well.  I have plans for a few approaches and have been thinking about settling on a fixed point format.  Long term, I want to be able to do more advanced data correlation in-kernel, up to and including FFTs.

It feels like ebpf in general needs work on the same thing, and I've found relatively little out there so far.  Ideally I'd hope that something could be done to benefit the wider community on this issue.

**Comment by janet-campbell** [comment]
BTW, here's another pretty one... percent iowait per IOP, all math happening within bpftrace:

<img width="1412" alt="image" src="https://github.com/bpftrace/bpftrace/assets/171300685/517f4839-ee2d-4f0b-b36c-9d4ce499cea6">
